internal/node/node_repositories.bzl

```diff
@@ -156,27 +266,98 @@ export PATH="$NODE_PATH":$PATH
-          "https://github.com/yarnpkg/yarn/releases/download/v1.3.2/yarn-v1.3.2.tar.gz",
-      ],
-      stripPrefix = "yarn-v1.3.2",
-      sha256 = "6cfe82e530ef0837212f13e45c1565ba53f5199eec2527b85ecbcd88bf26821d"
+    url = [url.format(version, filename) for url in urls],
+    stripPrefix = strip_prefix,
+    sha256 = sha256,
   )
 
-_yarn_repo = repository_rule(
-    _yarn_impl,
-    attrs = { "package_json": attr.label_list() },
+  _prepare_yarn(repository_ctx)
+
+yarn_download = repository_rule(
+  _yarn_download_impl,
+  attrs = {
+    "version": attr.string(),
+    "filename": attr.string(),
+    "strip_prefix": attr.string(),
+    "sha256": attr.string(),
+    "urls": attr.string_list(default = ["https://github.com/yarnpkg/yarn/releases/download/v{}/{}"]),
+    "package_json": attr.label_list(),
+  },
+)
+
+def _yarn_local_impl(repository_ctx):
+  """Used to register a vendored NodeJS runtime repository, usually one checked into the repo.
+
+  Repository Rule Args:
+    name - A unique name for this tool repo. This should almost always be yarn if you want the tool to be used by
+    this rule set.
+
+    path - The local path to a pre-installed yarn tool.
+
+    package_json: a list of labels, which indicate the package.json files that the yarn label will point to.
+
+  Args:
+    repository_ctx: The repository rule context
+  """
+  repository_ctx.symlink(repository_ctx.attr.path + "/yarn" , "bin/yarn")
+  _prepare_yarn(repository_ctx)
+
+yarn_local = repository_rule(
+  _yarn_local_impl,
+  attrs = {
+    "path": attr.string(),
+    "package_json": attr.label_list(),
+  },
 )
 
-def node_repositories(package_json):
+def node_repositories(package_json, node_version=DEFAULT_NODE_VERSION, yarn_version=DEFAULT_YARN_VERSION):
   """To be run in user's WORKSPACE to install rules_nodejs dependencies.
 
-  When the rule executes, it downloads node, npm, and yarn.
-  We fetch a specific version of Node, to ensure builds are hermetic, and to allow developers to skip installation of node if the entire toolchain is built on Bazel.
+  This rule sets up node, npm, and yarn.
+  
+  The versions of these tools can be specified in one of three ways:
+  - Normal Usage:
+    Specify no explicit versions. This will download and use the latest NodeJS & Yarn that were available when the
+    version of rules_nodejs you're using was released.
+  - Forced version(s):
+    You can select the version of NodeJS and/or Yarn to download & use by specifying it when you call node_repositories,
+    but you must use a value that matches a known version.
+  - Using a custom version:
+    You can call node_download_runtime or node_local_runtime to setup a particular NodeJS version or 
+    yarn_download or yarn_local to setup a particular Yarn version before calling node_repositories.
+    The "local" versions of these rules enable a repo to use vendored copies of tools rather than requiring a download.
```

Comment:
probably easier to understand if it's documented as a fourth way

Reply:
Done.